### PR TITLE
feat(#8841): Support multiplatform images

### PR DESCRIPTION
**Pull request by 1yuv**
# Description
Build multi-platform images
#8841 

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs
If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8841-multiplatform-images/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8841-multiplatform-images/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8841-multiplatform-images/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by 1yuv** [comment]
Working demo of external contributor: [this build](https://github.com/medic/cht-core/actions/runs/8195386118) and [internal contributor](https://github.com/medic/cht-core/actions/runs/8195026192)

**Review by mrjones-plip — CHANGES_REQUESTED** [review]
yay!!! This has been so long in the coming, it's just a real joy to finally see it in the form of PR - nice work!

tl;dr - I _think_ we may need to change  `const BUILD_PLATFORMS = ['linux/amd64', 'linux/arm64'];` to have `/v8` at the end like this: `const BUILD_PLATFORMS = ['linux/amd64', 'linux/arm64/v8'];`

[This SO post](https://stackoverflow.com/questions/70819028/relation-between-linux-arm64-and-linux-arm64-v8-are-these-aliases-for-each-othe) talks about it, but that talks about `linux/arm64/v8` normalizes to `linux/arm64` - but what about the other way around?

read below for my logic

-----

I'm doing some more research, but in my [earlier tests](https://medic.slack.com/archives/C02FMC810UE/p1708135273167069?thread_ts=1708134149.296279&cid=C02FMC810UE) I noted:

>  I do note that it's not `arm64` AND `amd64`, just `amd64`

This is because I tested the `cht-sentinel:4.6.0-8841-multiplatform-images` image.  Running a complete test, there's some more disparities, specifically that both sentinel and API are only `amd64`.  They're both based on `alpine:3.19` upstream.  I'm running tests with, sentinel, for example:

```
docker run --rm mplatform/mquery public.ecr.aws/medic/cht-sentinel:4.6.0-8841-multiplatform-images
```

Running an exhaustive test we see that none of the upstream images support `arm64` and they all support `arm64/v8`.  All CHT images are the `4.6.0-8841-multiplatform-images` tag:

| Image | `linux/amd64` | `linux/arm64` | `linux/arm64/v8` |
| -- | -- | -- | -- |
| cht-sentinel | Y | **N** | N |
| cht-couchdb | Y | Y | N |
| cht-nginx | Y | Y | N |
| cht-haproxy-healthcheck | Y | Y | N |
| cht-haproxy | Y | Y | N |
| cht-api | Y | **N** | N |
|  |  |  |  |
| alpine:3.19 (sentinel) | Y | N | Y |
| couchdb:3.3.3 (couchdb) | Y | N | Y |
| nginx:1.25.1-alpine (nginx) | Y | N | Y |
| python:3.10.13-alpine (haproxy-healthcheck) | Y | N | Y |
| haproxy:2.6 (haproxy) | Y | N | Y |
| alpine:3.19 (api) | Y | N | Y |

**Comment by mrjones-plip** [comment]
thanks @1yuv !  You know how long it takes to build new images after a commit/push based on your earlier efforts?

**Comment by 1yuv** [comment]
Hi @mrjones-plip , thanks for extensive test and details on the image and platforms they support. After updating build platfroms to `linux/arm64/v8`, I can see all medic images now support both amd64 and arm64. 

Medic Images | linux/amd64 | linux/arm64
-- | -- | -- 
cht-sentinel | Y | Y 
cht-couchdb | Y | Y 
cht-nginx | Y | Y 
cht-haproxy-healthcheck | Y | Y
cht-haproxy | Y | Y
cht-api | Y | Y

**Comment by mrjones-plip** [comment]
Awesome work @1yuv ! I was hoping we'd bet `linux/arm64/v8` out of this and not just `linux/arm64` :thinking:  Still, I'll approve this as is - I believe it's good to go. My local tests on Linux show no regressions, which is great.

That said I think we should get at least one other engineer to sign off on this before merging to `master`.

@dianabarsan or @nydr - any thoughts or concerns here? I kinda wanna get another set of eyes on this given that the new images can have far reaching implications if something goes off the rails. Maybe you can let us know before next Tue (Mar 12th) so we can keep this PR moving?

NB - I'll be out starting Mar 11th, back in the office Mar 18th.

**Comment by 1yuv** [comment]
Regarding time taken for building and publishing image, given we're building 2 platform images for 6 services, app compilation time has gone up by nearly 1 minute 30 seconds, specially `publish for testing` where we're building multi-platform images. This is in comparison with recent [single platform build action ](https://github.com/medic/cht-core/actions/runs/8136292579).

However, the overall build and test time is unaffected or reduced in comparison with some previous builds since we saved time in [Publishing branch build](https://github.com/medic/cht-core/actions/runs/8204273374/job/22443084584) by using [regctl](https://github.com/regclient/regclient) which detects pre-existing layers and prevent unnecessary i/o and has resulted in faster publishing of images compared to using `docker pull, tag, and push`.

**Review by dianabarsan — APPROVED** [review]
Cool. Thanks for getting this done, @1yuv !
